### PR TITLE
[Core] Consensus Mechanism (block rewards, slashing/jailing, staking, and delegation)

### PR DESCRIPTION
**Pull request by revnode**
Added 
 - Slashing Module
 - Distribution Module 
 
  Hooks for slashing and Distribution to Staking

### CODE REVIEW

**Comment by utx0** [comment]
Hey @revnode whats the status on this bro?

**Comment by revnode** [comment]
@utx0 Will ping you once it's ready. Should be today or tomorrow.

**Comment by revnode** [comment]
@utx0 Finished adding stuff. Will add unit tests next and make sure all is good, then we can review/merge.

**Comment by revnode** [comment]
Unit tests added.

File: app/app_test.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+	setGenesis(app)
+
+	// Making a new app object with the db, so that initchain hasn't been called
+	newApp := NewInitApp(log.NewTMLogger(log.NewSyncWriter(os.Stdout)), db, nil, true, 0)
+	_, _, err := newApp.ExportAppStateAndValidators(false, []string{})
+	require.NoError(t, err, "ExportAppStateAndValidators should not have an error")
+}
+
+// ensure that black listed addresses are properly set in bank keeper
+func TestBlackListedAddrs(t *testing.T) {
```
Comment (utx0) [inline app/app_test.go:29]: Hey mate, can you explain what the point of this blacklistedAddrs test is for? And what requirement are you trying to for-fill here?

File: app/export.go [hunk 2]
```diff
@@ -56,35 +57,71 @@ func (app *NewApp) prepForZeroHeightGenesis(ctx sdk.Context, jailWhiteList []str
 		whiteListMap[addr] = true
 	}
 
+	/* Handle fee distribution state. */
```
Comment (utx0) [inline app/export.go:60]: Can we add some tests around this export data code. Please?

**Review by utx0 — CHANGES_REQUESTED** [review]
Test still aren't passing and a few minor changes needed.

File: app/app_test.go (hunk 1 above)
Comment (revnode) [reply app/app_test.go:29]: Tests were a direct lift from Gaia. Free stuff and all that. If we want to modify them/have our own, we should come up with a list of things we want to test for and I can write them out. In terms of what's there, TestExport makes sense to me, but I agree, we don't need TestBlackListedAddrs. Is there anything you want me to add in terms of unit testing? PR is for stacking/slashing but we can add other stuff too.

File: app/export.go (hunk 2 above)
Comment (revnode) [reply app/export.go:60]: Yup, of course. See above. We should add more stuff too.

File: app/app_test.go (hunk 1 above)
Comment (Jedi2002) [reply app/app_test.go:26]: I just kept this test , it doesnt fail 
We dont have blacklsited address as such ,but there is no harm in having a test case ?

File: app/export.go (hunk 2 above)
Comment (Jedi2002) [reply app/export.go:60]: added a shell script to check state is exported properly .

**Comment by Jedi2002** [comment]
Hi @intl-man  , This is the pr I mentioned , which has the docker issue .
Have a look when you can .

**Comment by utx0** [comment]
> Hi @intl-man , This is the pr I mentioned , which has the docker issue .
> Have a look when you can .

I'm re-running the tests again as these sometime fail due to random issues

**Comment by Jedi2002** [comment]
> > Hi @intl-man , This is the pr I mentioned , which has the docker issue .
> > Have a look when you can .
> 
> I'm re-running the tests again as these sometime fail due to random issues

I did try running a couple of times , it gives the same error

File: app/app.go [hunk 3]
```diff
@@ -114,6 +121,8 @@ func NewInitApp(
 		oracle.StoreKey,
 		ethbridge.StoreKey,
 		clp.StoreKey,
+		distr.StoreKey,
+		slashing.StoreKey,
```
Comment (ElliotFriedman) [inline app/app.go:125]: Should we implement slashing for peggy?

Comment (utx0) [reply app/app.go:125]: What do you mean? This is for the sifchain? and there will effort both at this point.

**Review by banshee — CHANGES_REQUESTED** [review]
There's a panic in a transaction.  Take a look at sifchainstxs.txt:

```
height: 11
txhash: 854F35EBC52EEA2335F30591743F32195A54FC8E28E8E96C24360C888494394E
codespace: undefined
code: 111222
data: ""
rawlog: panic
logs: []
info: ""
gaswanted: 300000
gasused: 43836
tx:
  msg:
  - ethereum_chain_id: 5777
    bridge_registry_contract_address:
    - 242
    - 4
    - 164
    - 239
    - 8
    - 47
    - 92
    - 4
    - 187
    - 137
    - 247
    - 213
    - 230
    - 86
    - 139
    - 121
    - 96
    - 150
    - 115
    - 90
    nonce: 1
    symbol: eth
    token_contract_address:
    - 0
    - 0
    - 0
    - 0
    - 0
    - 0
    - 0
    - 0
    - 0
    - 0
    - 0
    - 0
    - 0
    - 0
    - 0
    - 0
    - 0
    - 0
    - 0
    - 0
    ethereum_sender:
    - 98
    - 115
    - 6
    - 9
    - 10
    - 186
    - 179
    - 166
    - 225
    - 64
    - 14
    - 147
    - 69
    - 188
    - 96
    - 199
    - 138
    - 139
    - 239
    - 87
    cosmos_receiver: sif1mu5cy44zykp3kjg7us4ewqhle8gykj2zfx7wtp
    validator_address: sifvaloper1mu5cy44zykp3kjg7us4ewqhle8gykj2zqykyt3
    amount: "10000000000000000000"
    claim_type: 0
  fee:
    amount: []
    gas: 300000
  signatures:
  - |
    pubkey: sifpub1addwnpepqdps0upf2lg5nrmjeqj6havensn6xg4dfnxzz8k9p093tf87flxxs6tz6jv
    signature: !!binary |
      X6/1Jxfi8p4mcIslvUfCtliLbJd1mKeSdjQE0Dmpkf9FaF9erW/0IFdwuaJc9ApZYacfDS
      ZQNTTRi/kpEp88uw==
  memo: ""
timestamp: "2020-12-22T01:50:58Z"

```

That's the output of running sifnodecli queries on the transactions in the log.


[datafiles.12-21-17-51-14.tar.gz](https://github.com/Sifchain/sifnode/files/5727365/datafiles.12-21-17-51-14.tar.gz)

**Review by utx0 — APPROVED** [review]

**Pull request merged**